### ISSUE TITLE
Twitter won't load - stuck in endless refresh prompt

### ISSUE THREAD

**Issue by dralley** [servo/servo#26032]
No error or explicit failure.  Clicking the refresh text refreshes the page but it just asks to refresh again afterwards, endlessly.

![Screenshot from 2020-03-25 10-03-35](https://user-images.githubusercontent.com/2751996/77544865-fe071380-6e7f-11ea-928f-6ca2f012076b.png)

**Comment by dralley** [comment]
re: https://github.com/servo/servo/issues/18818

**Comment by CYBAI** [comment]
Thanks to the `--load-script-source` option! We're able to debug twitter locally now!

Finally figured out why twitter can't be rendered correctly in Servo now.

Following is how I tried to debug it and the root cause in the end.

- [Debugging Process](#debugging-process)
- [Find the root cause](#find-the-root-case)
- [TL;DR](#tl-dr)

---

## Debugging Process

This is currently how twitter is rendered in Servo.

<img width="1136" alt="image" src="https://user-images.githubusercontent.com/6782666/83710740-603c7b80-a65c-11ea-8c26-92a7a5d3e823.png">

The twitter error message says, `Something went wrong, but don't fret - it's not your fault`.
With grepping from `unminified-js` folder, we can know the message is defined in `abs.twimg.com/responsive-web/web/i18n-rweb/en.f1d67e74.js` and it's equal to `aaf2506f`.

<img width="824" alt="image" src="https://user-images.githubusercontent.com/6782666/83710926-d5a84c00-a65c-11ea-96d4-749a86bdf1f5.png">

With grepping `aaf2506f`, we can know it's the `dJGd` react component because it has `componentDidCatch` for showing the error message.

<img width="1275" alt="image" src="https://user-images.githubusercontent.com/6782666/83711141-57987500-a65d-11ea-9ea5-ca66c7a0c1a7.png">

So, I added `console.log(t.componentStack)` into the `componentDidCatch` function; then, it shows that the error is from `circle` element inside `ActivityIndicator`.

<details>

<summary>Stacktrace of the react component</summary>

```
    in circle
    in svg
    in div
    in View
    in div
    in View
    in ActivityIndicator
    in @twitter/ActivityIndicator
    in div
    in View
    in t
    in n
    in div
    in View
    in div
    in View
    in div
    in View
    in t
    in div
    in View
    in @twitter/Responsive
    in @twitter/Container
    in div
    in View
    in header
    in View
    in t
    in i
    in _
    in routerWithMatchPropRemoved(Connect(i))
    in withRouter(routerWithMatchPropRemoved(Connect(i)))
    in div
    in View
    in @twitter/Responsive
    in t
    in t
    in u
    in ForwardRef
    in ForwardRef
    in t
    in t
    in t
    in _
    in t
    in u
    in ForwardRef
    in ForwardRef
    in t
    in Unknown
    in t
    in t
    in r
    in routerWithMatchPropRemoved(r)
    in withRouter(routerWithMatchPropRemoved(r))
    in t
    in t
    in t
    in _
    in s
    in t
    in div
    in View
    in div
    in View
    in r
```

</details>

Digging deeper into the `ActivityIndicator` component, we can find it's the `tyyN` react component in `abs.twimg.com/responsive-web/web/vendors~main.19622e04.js`. The `circle` elements and `svg` element are there.

<img width="1022" alt="image" src="https://user-images.githubusercontent.com/6782666/83711422-1bb1df80-a65e-11ea-9da9-7e5f933d94c2.png">

I tried several ways to see if twitter can be rendered correctly and following 3 kinds work fine to render a broken layout twitter

1. Change the `svg` to `div`
1. Remove the `circle` components
1. Remove `style` props of `circle`

The 3rd one of removing `style` just made me notice if it's related to `style`.

Then, I tried to create a minimized test case on codesandbox ([here](https://tsdxi.csb.app/)).

Luckily and thankfully, codesandbox uses a dev build of react and it will show the full stacktrace for why it fails.

<img width="1136" alt="image" src="https://user-images.githubusercontent.com/6782666/83718483-1493cd00-a670-11ea-96ee-675f9b1bbe7b.png">

---

## Find the root cause

https://github.com/facebook/react/blob/4c7036e807fa18a3e21a5182983c7c0f05c5936e/packages/react-dom/src/shared/CSSPropertyOperations.js#L52-L85

This is the `setValueForStyles` function in `react-dom` library.

It gets [`style`](https://drafts.csswg.org/cssom/#dom-elementcssinlinestyle-style) from the node (an element) and will set the style value to the style.

However, we don't support `style` for `SVGElement` yet so it just throws error while setting a value to `undefined` in [line 82](https://github.com/facebook/react/blob/4c7036e807fa18a3e21a5182983c7c0f05c5936e/packages/react-dom/src/shared/CSSPropertyOperations.js#L82).

---

## TL;DR

We don't support [`style`](https://drafts.csswg.org/cssom/#dom-elementcssinlinestyle-style) for `SVGElement` yet. React will use it to mutate styles so it will crash in Servo currently due to assigning a value to `undefined`.

**Comment by atouchet** [comment]
This still isn't working for me in the latest nightly. Am I doing something wrong?

**Comment by CYBAI** [comment]
The issue is related to SVG so you will need to enable SVG to make it work.

```
./mach run --pref dom.svg.enabled https://twitter.com
```

**Comment by atouchet** [comment]
Ah, that does appear to get it to load. The site has major visual issues when it loads though. I don't remember it being nearly that bad previously.
